### ISSUE TITLE
feat(storage): add in memory download & from memory upload

### ISSUE THREAD

**Issue by mamillastre** [capawesome-team/capacitor-firebase#665]
### Plugin(s)

- [ ] Analytics
- [ ] App
- [ ] App Check
- [ ] Authentication
- [ ] Crashlytics
- [ ] Cloud Firestore
- [ ] Cloud Functions
- [ ] Cloud Messaging
- [X] Cloud Storage
- [ ] Performance
- [ ] Remote Config

### Current problem

The plugin allows to download/upload files to/from local file (except for the web upload that support blob).

### Preferred solution

Firebase can also download/upload files to/from memory data.
We can add a "getData()" method and add a "data" param to the "uploadFile()".
I think the best with the plugin constraints is to manipulate base64 data like the FileSystem plugin.

The getData() method will also allow to have an alternative to the "public" link returned by getDownloadUrl().

### Additional context

Upload:
- Web: https://firebase.google.com/docs/storage/web/upload-files#upload_from_a_string
- Android: https://firebase.google.com/docs/storage/android/upload-files#upload_from_data_in_memory
- iOS: https://firebase.google.com/docs/storage/ios/upload-files#upload_from_data_in_memory

Download:
- Web: https://firebase.google.com/docs/storage/web/download-files#download_data_directly_from_the_sdk
- Android: https://firebase.google.com/docs/storage/android/download-files#download_in_memory
- iOS: https://firebase.google.com/docs/storage/ios/download-files#download_in_memory

### Before submitting

- [X] I have read and followed the [feature request guidelines](https://capawesome.io/contributing/feature-requests/).
- [X] I have attached links to possibly related issues and discussions.

**Comment by robingenz** [comment]
I like the idea, but using base64 leads to an out-of-memory exception with "large" files. For this reason, I always recommend interacting directly with the native file system (see this [blog post](https://capawesome.io/blog/the-file-handling-guide-for-capacitor/)). I'm not sure whether it's a good idea to make this option available to Capacitor developers. But feel free to create a PR and I'll take a look at it.

**Comment by mamillastre** [comment]
Good point.
It can be useful for small files and to avoid the security issue of the download URL when a file is protected by a rule.
I stand by this feature for now. I may be working on this in the future.
Thank you.

**Comment by robingenz** [comment]
>  It can be useful for small files and to avoid the security issue of the download URL when a file is protected by a rule.

Yes, that's a good point. I would probably merge the feature and wait and see what the feedback from the community is like.

**Comment by mamillastre** [comment]
@robingenz, maybe an alternative solution to fix the security issue is to use the "download to a local file" functionality available on Android & iOS.

Android: https://firebase.google.com/docs/storage/android/download-files#download_to_a_local_file
iOS: https://firebase.google.com/docs/storage/ios/download-files#download_to_a_local_file

We can have a "getFile()" method that has a call signature similar to the "Filesystem.writeFile()" method's interfaces (WriteFileOptions & WriteFileResult).

What do you think of this solution ?

**Comment by robingenz** [comment]
That is much better. However, I would probably prefer to call the method `downloadFile(...)` so that it matches `uploadFile(...)`.

> We can have a "getFile()" method that has a call signature similar to the "Filesystem.writeFile()" method's interfaces (WriteFileOptions & WriteFileResult).

I suggest the following for the call signature:

```ts
export interface DownloadFileOptions {
  /**
   * The full path to the file to download, including the file name.
   * 
   * @since 6.2.0
   */
  remotePath: string;
  /**
   * The uri to the local file, including the file name.
   * 
   * @since 6.2.0
   */
  localPath: string;
}
```
`data` is not required. Instead of using `directory` and `path`, I would rather require a uri (called `localPath` here) directly. Otherwise you always have to keep the types synchronized with the official Capacitor Filesystem plugin. And for now, I would not want to add a `recursive` option to keep it simple.

@mamillastre What do you think?

**Comment by mamillastre** [comment]
Seems perfect !
With the help of the Filesystem plugin, we can manage the permission, create the directory (`mkdir()`) and get the uri (`getUri()`).

To be consistent with the Storage plugin and the Filesystem plugin, we can also have:
```ts
export interface DownloadFileOptions {
  /**
   * The full path to the file to download, including the file name.
   * 
   * @since 6.2.0
   */
  path: string;
  /**
   * The uri to the local file, including the file name.
   * 
   * @since 6.2.0
   */
  uri: string;
}
```

I will update this feature request.

For now, I do not have the time to work on this. But maybe later.

**Comment by robingenz** [comment]
You're right, that's better.
